### DISCUSSION TITLE
Seeking Advice: Improving Detection When Animal Is Absent in Most Frames

### DISCUSSION THREAD

**Discussion by skyshi0** [talmolab/sleap#2445] [Help!] [answered]
Hi SLEAP team and community,

I’m seeking advice on improving my model’s ability to avoid **false positives** when my animal is not present in most frames.
### Setup:

- Video resolution: 1080×1080 px, ~1px per mm
- Animal size:18-25 px length
- Animal presence: ~15% of total frames

The bottom-up model performs okay when the target is present and visible. However, since I want to use the detection timestamps for downstream behavioral analysis, I need the model to also correctly identify when the animal is absent (i.e., predict no instances).

### Problem
Initially, I trained the model based on ~1200 labeled frames only when the target is present. The bottom-up model seems to have fewer false positive frames compared to top-down and single models when I ran for an entire video(visually inspected, I haven’t systematically quantified the false positive frames). So I was thinking whether I could improve its performance by adding more ‘invisible frames’(meaning add those frames as labeled frames but set all the nodes to invisible),  but after I added 133 ‘invisible frames’, the bottom-up model seems to not improve and still identified random locations as the target.

### Question
Do you have any advice on how to improve model training when:(1)The target is very small (~19 px),(2)The target only appears in a small subset of frames (~15%), (3)And I want to minimize false positives when the target is absent?

I’m wondering:
1. Should I increase or decrease the ratio of “invisible frames” in the training dataset or does it matter?
2. . Are there recommended data augmentation or preprocessing steps for such small targets?
3.  Is the resolution too low to identify the target? I know it doesn't meet the criteria for mouse [~2mm per pixel](https://www.nature.com/articles/s41592-022-01426-1#:~:text=The%20%E2%80%98mice_of%E2%80%99%20dataset,at%2080%E2%80%89FPS.), but I wonder if it can still be improved since I don't need the fine features of the target.
<img width="1621" height="931" alt="image" src="https://github.com/user-attachments/assets/3155c754-5c9d-4470-b496-b9659586042a" />
<img width="313" height="295" alt="image" src="https://github.com/user-attachments/assets/8048c8d8-95a9-4ba9-a27b-83ab79a23a3e" />
<img width="305" height="311" alt="image" src="https://github.com/user-attachments/assets/2152aab0-0724-4ec4-8333-0903ec9f1490" />

Any suggestions or examples from similar projects would be greatly appreciated!
Thanks in advance for your time and insights!

**Comment by gitttt-1234** [comment, marked as answer]
Hi @skyshi0 !

Apologies for the delay!

The core issue is that **setting all nodes to "invisible" does not create a negative training example**. To teach the model "nothing is here," here's what we'd recommend (this was a recent update to our training pipeline, I would recommend to update to the latest SLEAP version v1.6.3. We previously didn't have this feature, so even if you had added negative examples, the training pipeline would skip them as it only used frames containing labeled instances):

1. Use negative/empty frames

Instead of adding instances with invisible nodes, label frames with **no instances at all**. In SLEAP 1.6+, you can do this by:
- Navigating to a frame where the animal is absent
- Making sure there are no instances on that frame (delete any if present)
- The frame should show up as a labeled frame with 0 instances

To do this programmatically for a batch of frames:

```python
import sleap_io as sio

labels = sio.load_slp("your_project.slp")

# Add empty negative frames (example: every 100th frame from video 0)
video = labels.videos[0]
for frame_idx in range(0, 5000, 100):
    lf = sio.LabeledFrame(video=video, frame_idx=frame_idx)
    lf.is_negative = True
    labels.labeled_frames.append(lf)

sio.save_slp(labels, "your_project_with_negatives.slp")
```

2. Enable negative frames in training

If you're using CLI training with sleap-nn, add this to your training config YAML:

```yaml
data_config:
  use_negative_frames: true
  negative_loss_weight: 1.0  # increase to 2.0+ if false positives persist
```

3. Adjust the ratio

With ~85% of real-world frames being empty, having roughly 30-50% negative frames in your training set gives the model a strong signal. Your 133 invisible-node frames out of ~1333 was only ~10% *and* the wrong type — so it's no surprise it didn't help.

4. Post-inference score thresholding

Even without retraining, you can filter out low-confidence detections after inference. Predictions with low confidence scores are often the false positives. You can threshold these either by passing `--peak_threshold` on sleap-nn track CLI or on the predicted slp file with the following script:

```python
import sleap_io as sio

labels = sio.load_slp("predictions.slp")
for lf in labels:
    lf.instances = [inst for inst in lf.instances if inst.score > 0.3]
sio.save_slp(labels, "predictions_filtered.slp")
```

Adjust the `0.3` threshold based on your data. You might want to find the sweet spot that removes noise peaks without dropping real detections.

### 5. Small target tips

Top-down is usually effective for animals that are too small relative to the frame, as we crop around the animal before passing to the centered-instance model where the keypoints are predicted.

Let us know if you have any questions!

Related thread: https://github.com/talmolab/sleap/issues/640

Thanks,

Divya

**Reply by skyshi0** [reply]
Hi Divya,
Thank you so much! These steps were exactly what I was testing:

1. invisible node: I personally prefer to add the frame to labeled frame->add an instance ->turn all the nodes to invisible; then use sleap-io to find all the labeled frames that have an instance but no visible nodes and flag them to negative frames, the way @davorvr did before. This is because at the beginning I usually used the random sampling methods to generate suggested frames but didn't end up labeling them, and I don't want to treat those frames as negative frames. Do you have plans for designing a button or panel in the GUI to do so?
2. So far I added ~10% negative frames (443 negative frames) along with 3924 positive frames; the false negative frames were somehow higher than the normal model, but perhaps I could improve the model further by adding more negative frames as you suggested to ~30%.
3. Yes. I found it effectively separate them (attached a picture) of finding the threshold to separate them, but I had some hard frames that I would want to preserve if possible but discriminate them from the negative frames(unfortunately interpolation couldn't help me with this so I need to improve the tracking itself).
4. Top-down model is my favorite right now. It's the easiest to train and find the best threshold, I'm now playing around with 4 different conditions: with/without negative frames and with/without brightness/contrast argumentation.

<img width="507" height="145" alt="image" src="https://github.com/user-attachments/assets/0a4d592f-ffea-4bbc-8c54-ced4e6494e7e" />
<img width="1989" height="396" alt="image" src="https://github.com/user-attachments/assets/fbe05021-1310-4f07-b651-a9c370074fd8" />
